File: pymd/backend/app/schemas/document.py
"""Document schemas"""
from datetime import datetime
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, Field


class DocumentBase(BaseModel):
    """Base document schema"""

    title: str = Field(..., min_length=1, max_length=255)
    content: str = Field(default="")
    render_format: str = Field(default="html", pattern="^(html|markdown)$")


class DocumentCreate(DocumentBase):
    """Document creation schema"""

    pass


class DocumentUpdate(BaseModel):
    """Document update schema"""

    title: Optional[str] = Field(None, min_length=1, max_length=255)
    content: Optional[str] = None
    render_format: Optional[str] = Field(None, pattern="^(html|markdown)$")


class DocumentResponse(DocumentBase):
    """Document response schema"""

    id: UUID
    owner_id: UUID
    is_deleted: bool
    created_at: datetime
    updated_at: datetime
    last_accessed_at: datetime

    model_config = {"from_attributes": True}


class DocumentListResponse(BaseModel):
    """Document list response with pagination"""

    documents: list[DocumentResponse]
    total: int
    page: int
    page_size: int
    has_more: bool

File: pymd_frontend/types/document.ts
/**
 * Document type definitions
 */

export interface Document {
  id: string;
  owner_id: string;
  title: string;
  content: string;
  render_format: 'html' | 'markdown';
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  last_accessed_at: string;
}

export interface DocumentCreate {
  title: string;
  content?: string;
  render_format?: 'html' | 'markdown';
}

export interface DocumentUpdate {
  title?: string;
  content?: string;
  render_format?: 'html' | 'markdown';
}

export interface DocumentListResponse {
  documents: Document[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface DocumentListParams {
  page?: number;
  page_size?: number;
  search?: string;
  sort_by?: 'created_at' | 'updated_at' | 'title';
  sort_order?: 'asc' | 'desc';
}

export interface RenderRequest {
  content: string;
  format?: 'html' | 'markdown';
}

export interface RenderResponse {
  rendered: string;
  cached: boolean;
}

export interface ValidationError {
  line: number;
  column: number;
  message: string;
  severity: 'error' | 'warning';
}

export interface ValidationResponse {
  valid: boolean;
  errors: ValidationError[];
}
